USE instruction failing in some cases

# USE instruction failing in some cases

Hello ProxySQL team !

## Description
Some of our databases were inaccessible with ProxySQL 2.2.1.  
It seems like ProxySQL 2.2.1 has some issues with databases containing '-' and a number next to it in their name.  
After rolling back ProxySQL to 2.2.0, everything was working properly.

## ProxySQL version
 - `ProxySQL version 2.2.1-39-g837a720, codename Truls` amd64 on Debian
 - `ProxySQL version 2.2.1-39-g837a720, codename Truls` arm64 with the official image from DockerHub

## Steps to reproduce
The ProxySQL configuration is the one provided on DockerHub.  
This bug has been reproduced with the following `docker-compose`:

```YAML
version: '3'

services:
  proxysql:
    image: proxysql/proxysql:2.2.1
    ports:
      - 6033:6033
      - 6032:6032
    environment:
      - MONITOR_CONFIG_CHANGE=true
    volumes:
      - ./proxysql.cnf:/etc/proxysql.cnf

  mariadb0:
    image: mariadb:10.4.21-focal
    environment:
      - MYSQL_ROOT_PASSWORD=toor
      - MARIADB_USER=monitor
      - MARIADB_PASSWORD=monitor
    ports:
      - 3306:3306
```

 - Create the following databases on the MariaDB instance
   - `test1`
   - `test-1`
   - `test-1es`
   - `test-test`
 - Connect to ProxySQL admin port and execute the following statements :
```SQL
INSERT INTO `mysql_servers` (`hostgroup_id`, `hostname`, `port`, `comment`) VALUES ('0', 'mariadb0', '3306', 'default instance');
INSERT INTO `mysql_replication_hostgroups` (`writer_hostgroup`, `reader_hostgroup`, `comment`) VALUES ('0', '1', 'default');
INSERT INTO `mysql_users` (`username`, `password`, `active`, `default_hostgroup`) VALUES ('root', 'toor', '1', '0');
INSERT INTO `mysql_query_rules` (`rule_id`, `active`, `flagOUT`, `apply`) VALUES ('0', '1', '0', '0');
INSERT INTO `mysql_query_rules_fast_routing` (`username`, `schemaname`, `flagIN`, `destination_hostgroup`, `comment`) VALUES ('', 'test1', '0', '0', ''), ('', 'test-1es', '0', '0', ''), ('', 'test-1', '0', '0', ''), ('', 'test-test', '0', '0', '');

LOAD MYSQL SERVERS TO RUNTIME;
LOAD MYSQL USERS TO RUNTIME;
LOAD MYSQL QUERY RULES TO RUNTIME;
```
 - Connect to ProxySQL data port and execute the following statements :
 ```SQL
USE `test1`;
SELECT 1; # should work properly
USE `test-test`;
SELECT 1; # should work properly
USE `test-1`;
SELECT 1; # should throw an error
USE `test-1es`;
SELECT 1; # should throw an error
```

## ProxySQL error log
```
2021-09-07 13:21:25 mysql_connection.cpp:1006:handler(): [ERROR] Failed to mysql_real_connect() on 0:mariadb0:3306 , FD (Conn:31 , MyDS:31) , 1049: Unknown database 'test-?'.
2021-09-07 13:21:26 mysql_connection.cpp:1587:handler(): [ERROR] Failed INITDB: Unknown database 'test-?'
2021-09-07 13:21:26 MySQL_Session.cpp:2594:handler_again___status_CHANGING_SCHEMA(): [WARNING] Error during INIT_DB: 1049, Unknown database 'test-?'
2021-09-07 13:21:26 MySQL_Session.cpp:149:kill_query_thread(): [WARNING] KILL CONNECTION 51 on mariadb0:3306

2021-09-07 13:21:28 mysql_connection.cpp:1006:handler(): [ERROR] Failed to mysql_real_connect() on 0:mariadb0:3306 , FD (Conn:31 , MyDS:31) , 1049: Unknown database 'test-1s'.
2021-09-07 13:21:28 mysql_connection.cpp:1587:handler(): [ERROR] Failed INITDB: Unknown database 'test-1s'
2021-09-07 13:21:28 MySQL_Session.cpp:2594:handler_again___status_CHANGING_SCHEMA(): [WARNING] Error during INIT_DB: 1049, Unknown database 'test-1s'
2021-09-07 13:21:28 MySQL_Session.cpp:149:kill_query_thread(): [WARNING] KILL CONNECTION 97 on mariadb0:3306
```

I hope this will help.

Have a nice day !

--- a/include/MySQL_Session.h
+++ b/include/MySQL_Session.h
@@ -90,7 +90,7 @@ class MySQL_Session
 	 *   But since it was change for handling 'USE' statements which are preceded by
 	 *   comments, it's called after 'QueryProcessor' has processed the query.
 	 */
-	void handler___status_WAITING_CLIENT_DATA___STATE_SLEEP___MYSQL_COM_QUERY_USE_DB(PtrSize_t *pkt, const char* query_digest);
+	void handler___status_WAITING_CLIENT_DATA___STATE_SLEEP___MYSQL_COM_QUERY_USE_DB(PtrSize_t *pkt);
 	void handler___status_WAITING_CLIENT_DATA___STATE_SLEEP___MYSQL_COM_PING(PtrSize_t *);
 
 	void handler___status_WAITING_CLIENT_DATA___STATE_SLEEP___MYSQL_COM_CHANGE_USER(PtrSize_t *, bool *);

--- a/lib/MySQL_Session.cpp
+++ b/lib/MySQL_Session.cpp
@@ -3592,8 +3592,18 @@ __get_pkts_from_client:
 									// For more context check issue: #3493.
 									// ===================================================
 									if (session_type != PROXYSQL_SESSION_CLICKHOUSE) {
-										if (strncasecmp((char *)"USE ",CurrentQuery.get_digest_text(),4)==0) {
-											handler___status_WAITING_CLIENT_DATA___STATE_SLEEP___MYSQL_COM_QUERY_USE_DB(&pkt, CurrentQuery.get_digest_text());
+										fprintf(stderr,"%s\n", CurrentQuery.get_digest_text());
+										const char *qd = CurrentQuery.get_digest_text();
+										if (
+											(strncasecmp((char *)"USE",qd,3)==0)
+											&&
+											(
+												(strncasecmp((char *)"USE ",qd,4)==0)
+												||
+												(strncasecmp((char *)"USE`",qd,4)==0)
+											)
+										) {
+											handler___status_WAITING_CLIENT_DATA___STATE_SLEEP___MYSQL_COM_QUERY_USE_DB(&pkt);
 
 											if (mirror == false) {
 												break;
@@ -5241,18 +5251,25 @@ void MySQL_Session::handler___status_WAITING_CLIENT_DATA___STATE_SLEEP___MYSQL_C
 
 // this function was introduced due to isseu #718
 // some application (like the one written in Perl) do not use COM_INIT_DB , but COM_QUERY with USE dbname
-void MySQL_Session::handler___status_WAITING_CLIENT_DATA___STATE_SLEEP___MYSQL_COM_QUERY_USE_DB(PtrSize_t *pkt, const char* query_digest) {
+void MySQL_Session::handler___status_WAITING_CLIENT_DATA___STATE_SLEEP___MYSQL_COM_QUERY_USE_DB(PtrSize_t *pkt) {
 	gtid_hid=-1;
 	proxy_debug(PROXY_DEBUG_MYSQL_COM, 5, "Got COM_QUERY with USE dbname\n");
 	if (session_type == PROXYSQL_SESSION_MYSQL) {
 		__sync_fetch_and_add(&MyHGM->status.frontend_use_db, 1);
-		char *schemaname=strndup(query_digest+4,strlen(query_digest)-4);
+		string nq=string((char *)pkt->ptr+sizeof(mysql_hdr)+1,pkt->size-sizeof(mysql_hdr)-1);
+		RE2::GlobalReplace(&nq,(char *)"(?U)/\\*.*\\*/",(char *)" ");
+		char *sn_tmp = (char *)nq.c_str();
+		while (sn_tmp < ( nq.c_str() + nq.length() - 4 ) && *sn_tmp == ' ')
+			sn_tmp++;
+		//char *schemaname=strdup(nq.c_str()+4);
+		char *schemaname=strdup(sn_tmp+3);
 		char *schemanameptr=trim_spaces_and_quotes_in_place(schemaname);
 		// handle cases like "USE `schemaname`
 		if(schemanameptr[0]=='`' && schemanameptr[strlen(schemanameptr)-1]=='`') {
 			schemanameptr[strlen(schemanameptr)-1]='\0';
 			schemanameptr++;
 		}
+		proxy_info("%s\n", schemanameptr);
 		client_myds->myconn->userinfo->set_schemaname(schemanameptr,strlen(schemanameptr));
 		free(schemaname);
 		if (mirror==false) {

--- a/lib/c_tokenizer.c
+++ b/lib/c_tokenizer.c
@@ -484,6 +484,11 @@ char *mysql_query_digest_and_first_comment(char *s, int _len, char **first_comme
 					}
 					cmd=0;
 				}
+				if (flag == 1 && prev_char == '*' && *s == '/') {
+					if (r != p_r && *p_r != ' ') { // not at the beginning, and previous char is not ' '
+						*p_r++ = ' ';
+					}
+				}
 				prev_char = ' ';
 				flag = 0;
 				s++;

--- a/test/tap/tests/reg_test_3493-USE_with_comment-t.cpp
+++ b/test/tap/tests/reg_test_3493-USE_with_comment-t.cpp
@@ -16,6 +16,7 @@
 
 #include <cstring>
 #include <vector>
+#include <utility>
 #include <string>
 #include <stdio.h>
 
@@ -37,6 +38,8 @@ void parse_result_json_column(MYSQL_RES *result, json& j) {
 	}
 }
 
+std::vector<std::pair<std::string,std::string>> db_query;
+
 int main(int argc, char** argv) {
 	CommandLine cl;
 
@@ -46,6 +49,24 @@ int main(int argc, char** argv) {
 	}
 
 	MYSQL* proxysql_mysql = mysql_init(NULL);
+
+	db_query.push_back(std::make_pair("reg_test_3493_use_comment", "/*+ placeholder_comment */ USE reg_test_3493_use_comment"));
+	db_query.push_back(std::make_pair("`reg_test_3493_use_comment-a1`", "USE /*+ placeholder_comment */ `reg_test_3493_use_comment-a1`"));
+	db_query.push_back(std::make_pair("reg_test_3493_use_comment_1", "  USE /*+ placeholder_comment */   `reg_test_3493_use_comment_1`"));
+	db_query.push_back(std::make_pair("reg_test_3493_use_comment_2", "USE/*+ placeholder_comment */ `reg_test_3493_use_comment_2`"));
+	db_query.push_back(std::make_pair("reg_test_3493_use_comment_3", "USE /*+ placeholder_comment */`reg_test_3493_use_comment_3`"));
+	db_query.push_back(std::make_pair("reg_test_3493_use_comment_4", "  USE /*+ placeholder_comment */   reg_test_3493_use_comment_4"));
+	db_query.push_back(std::make_pair("reg_test_3493_use_comment_5", "USE/*+ placeholder_comment */ reg_test_3493_use_comment_5"));
+	db_query.push_back(std::make_pair("reg_test_3493_use_comment_6", "USE /*+ placeholder_comment */reg_test_3493_use_comment_6"));
+	db_query.push_back(std::make_pair("`reg_test_3493_use_comment-1`", "  USE /*+ placeholder_comment */   `reg_test_3493_use_comment-1`"));
+	db_query.push_back(std::make_pair("`reg_test_3493_use_comment-2`", "USE/*+ placeholder_comment */ `reg_test_3493_use_comment-2`"));
+	db_query.push_back(std::make_pair("`reg_test_3493_use_comment-3`", "USE /*+ placeholder_comment */`reg_test_3493_use_comment-3`"));
+	db_query.push_back(std::make_pair("`reg_test_3493_use_comment-4`", "/*+ placeholder_comment */USE          `reg_test_3493_use_comment-4`"));
+	db_query.push_back(std::make_pair("`reg_test_3493_use_comment-5`", "USE/*+ placeholder_comment */`reg_test_3493_use_comment-5`"));
+	db_query.push_back(std::make_pair("`reg_test_3493_use_comment-6`", "/* comment */USE`reg_test_3493_use_comment-6`"));
+	db_query.push_back(std::make_pair("`reg_test_3493_use_comment-7`", "USE`reg_test_3493_use_comment-7`"));
+
+	plan(db_query.size());
 
 	if (
 		!mysql_real_connect(
@@ -60,46 +81,81 @@ int main(int argc, char** argv) {
 	}
 
 	// Prepare the DB for the test
-	MYSQL_QUERY(proxysql_mysql, "DROP DATABASE IF EXISTS reg_test_3493_use_comment");
-	MYSQL_QUERY(proxysql_mysql, "CREATE DATABASE reg_test_3493_use_comment");
+	for (std::vector<std::pair<std::string,std::string>>::iterator it = db_query.begin(); it != db_query.end() ; it++) {
+	//	MYSQL_QUERY(proxysql_mysql, "DROP DATABASE IF EXISTS reg_test_3493_use_comment");
+	//	MYSQL_QUERY(proxysql_mysql, "CREATE DATABASE reg_test_3493_use_comment");
+		std::string s = "";
+		s = "DROP DATABASE IF EXISTS " + it->first;
+		MYSQL_QUERY(proxysql_mysql, s.c_str());
+		s = "CREATE DATABASE " + it->first;
+		MYSQL_QUERY(proxysql_mysql, s.c_str());
+	}
 
-	int err = mysql_query(proxysql_mysql, "/*+ placeholder_comment */ USE reg_test_3493_use_comment");
-	if (err) {
-		diag(
-			"'USE' command failed with error code '%d' and error '%s'",
-			err, mysql_error(proxysql_mysql)
+	for (std::vector<std::pair<std::string,std::string>>::iterator it = db_query.begin(); it != db_query.end() ; it++) {
+		int i = 0;
+		int err = mysql_query(proxysql_mysql, it->second.c_str());
+		if (err) {
+			diag(
+				"'USE' command failed with error code '%d' and error '%s' for query: %s",
+				err, mysql_error(proxysql_mysql), it->second.c_str()
+			);
+			return EXIT_FAILURE;
+		}
+
+		// Perform the 'SELECT DATABASE()' query in a new backend connection, to
+		// verify that ProxySQL is properly tracking the previously performed 'USE'
+		// statement.
+		switch (i%5) {
+			case 0:
+				MYSQL_QUERY(proxysql_mysql, "/*+ ;create_new_connection=1 */ SELECT DATABASE()");
+				break;
+			case 1:
+				MYSQL_QUERY(proxysql_mysql, "/*+ ;create_new_connection=1 */SELECT DATABASE()");
+				break;
+			case 2:
+				MYSQL_QUERY(proxysql_mysql, "SELECT /*+ ;create_new_connection=1 */ DATABASE()");
+				break;
+			case 3:
+				MYSQL_QUERY(proxysql_mysql, "SELECT/*+ ;create_new_connection=1 */ DATABASE()");
+				break;
+			case 4:
+				MYSQL_QUERY(proxysql_mysql, "SELECT /*+ ;create_new_connection=1 */DATABASE()");
+				break;
+			default:
+				assert(0);
+		}
+		i++;
+		MYSQL_RES* result = mysql_store_result(proxysql_mysql);
+		if (result == nullptr) {
+			diag("Invalid 'MYSQL_RES' returned from 'SELECT DATABASE()'");
+			return EXIT_FAILURE;
+		}
+
+		MYSQL_ROW row = mysql_fetch_row(result);
+		if (row == nullptr) {
+			diag("Invalid 'MYSQL_ROW' returned from 'SELECT DATABASE()'");
+			return EXIT_FAILURE;
+		}
+
+		std::string database_name { row[0] };
+
+		if (it->first[0] == '`') {
+			database_name = "`" + database_name + "`";
+		}
+		ok(
+			database_name == it->first,
+			"Selected DB name should be equal to actual DB name: (Exp: '%s') == (Act: '%s')",
+			it->first.c_str(),
+			database_name.c_str()
 		);
-		return EXIT_FAILURE;
 	}
-
-	// Perform the 'SELECT DATABASE()' query in a new backend connection, to
-	// verify that ProxySQL is properly tracking the previously performed 'USE'
-	// statement.
-	MYSQL_QUERY(proxysql_mysql, "/*+ ;create_new_connection=1 */ SELECT DATABASE()");
-	MYSQL_RES* result = mysql_store_result(proxysql_mysql);
-	if (result == nullptr) {
-		diag("Invalid 'MYSQL_RES' returned from 'SELECT DATABASE()'");
-		return EXIT_FAILURE;
-	}
-
-	MYSQL_ROW row = mysql_fetch_row(result);
-	if (row == nullptr) {
-		diag("Invalid 'MYSQL_ROW' returned from 'SELECT DATABASE()'");
-		return EXIT_FAILURE;
-	}
-
-	std::string database_name { row[0] };
-
-	ok(
-		database_name == "reg_test_3493_use_comment",
-		"Selected DB name should be equal to actual DB name: (Exp: '%s') == (Act: '%s')",
-		"reg_test_3493_use_comment",
-		database_name.c_str()
-	);
 
 	// Drop created database
-	MYSQL_QUERY(proxysql_mysql, "DROP DATABASE IF EXISTS reg_test_3493_use_comment");
-
+	for (std::vector<std::pair<std::string,std::string>>::iterator it = db_query.begin(); it != db_query.end() ; it++) {
+		std::string s = "";
+		s = "DROP DATABASE IF EXISTS " + it->first;
+		MYSQL_QUERY(proxysql_mysql, s.c_str());
+	}
 	mysql_close(proxysql_mysql);
 
 	return exit_status();

